File: leetcode/Offer36.cpp
#include <cstdlib>

// Definition for a Node.
class Node {
public:
  int val;
  Node *left;
  Node *right;

  Node() {}

  Node(int _val) {
    val = _val;
    left = NULL;
    right = NULL;
  }

  Node(int _val, Node *_left, Node *_right) {
    val = _val;
    left = _left;
    right = _right;
  }
};

class Solution {
private:
  Node *cur;
  void dfs(Node *root) {
    if (root == NULL) {
      return;
    }
    dfs(root->left);
    cur->right = root;
    root->left = cur;
    cur = root;
    dfs(root->right);
  }

public:
  Node *treeToDoublyList(Node *root) {
    if (root == NULL) {
      return root;
    }
    Node head(0);
    cur = &head;
    dfs(root);
    cur->right = head.right;
    head.right->left = cur;
    return head.right;
  }
};

File: leetcode/Offer36.py
# Definition for a Node.
class Node:

    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def treeToDoublyList(self, root: Node) -> Node:

        def dfs(root: Node) -> Node:
            nonlocal cur
            if root == None:
                return None
            dfs(root.left)
            cur.right = root
            root.left = cur
            cur = root
            dfs(root.right)

        if not root:
            return root
        head = Node(0)
        cur = head
        dfs(root)
        head.right.left, cur.right = cur, head.right
        return head.right